--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -270,6 +_,11 @@
         org.purpurmc.purpur.PurpurConfig.registerCommands();
         */// Purpur end - Purpur config files // Purpur end - Migrate Setting to reintroduce end void rings
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
+        // Sakura start - sakura configuration files
+        sakuraConfigurations.initializeGlobalConfiguration(this.registryAccess());
+        sakuraConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        me.samsuik.sakura.command.SakuraCommands.registerCommands(this);
+        // Sakura end - sakura configuration files
 
         // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
