--- a/net/minecraft/world/entity/animal/golem/IronGolem.java
+++ b/net/minecraft/world/entity/animal/golem/IronGolem.java
@@ -60,6 +_,13 @@
     private @Nullable EntityReference<LivingEntity> persistentAngerTarget;
     private java.util.@Nullable UUID summoner; // Purpur - Summoner API
 
+    // Sakura start - configure iron golems taking fall damage
+    @Override
+    protected final boolean isFallDamageImmune() {
+        return !this.level().sakuraConfig().entity.ironGolemsTakeFalldamage && super.isFallDamageImmune();
+    }
+    // Sakura end - configure iron golems taking fall damage
+
     public IronGolem(final EntityType<? extends IronGolem> type, final Level level) {
         super(type, level);
     }
