--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -192,12 +_,32 @@
 
     @Override
     public int getTickDelay(LevelReader level) {
-        return level.getWorldBorder().world != null ? (level.dimensionType().ultraWarm() ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether) : (level.dimensionType().ultraWarm() ? 10 : 30); // Purpur - Make lava flow speed configurable
-    }
+        // Sakura start - configure fast nether lava
+        final boolean netherLava = level.dimensionType().ultraWarm()
+            && !(level instanceof Level gameLevel
+            && gameLevel.sakuraConfig().environment.disableFastNetherLava);
+
+        if (level.getWorldBorder().world == null) {
+            return netherLava ? 10 : 30;
+        } else {
+            return netherLava
+                ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether // Purpur - Make lava flow speed configurable
+                : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether; // Purpur - Make lava flow speed configurable
+        }
+        // Sakura end - configure fast nether lava
+    }
+
+    // Sakura start - lava flow speed api
+    @Override
+    public final int getTickDelay(final Level world, final BlockPos pos) {
+        final int flowSpeed = world.localConfig().at(pos).lavaFlowSpeed;
+        return flowSpeed >= 0 ? flowSpeed : this.getTickDelay(world);
+    }
+    // Sakura end - lava flow speed api
 
     @Override
     public int getSpreadDelay(Level level, BlockPos pos, FluidState currentState, FluidState newState) {
-        int tickDelay = this.getTickDelay(level);
+        int tickDelay = this.getTickDelay(level, pos); // Sakura - lava flow speed api
         if (!currentState.isEmpty()
             && !newState.isEmpty()
             && !currentState.getValue(FALLING)
