--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -446,7 +_,7 @@
         return shouldConnectTo(state, null);
     }
 
-    protected static boolean shouldConnectTo(BlockState state, @Nullable Direction direction) {
+    public static boolean shouldConnectTo(BlockState state, @Nullable Direction direction) {
         if (state.is(Blocks.REDSTONE_WIRE)) {
             return true;
         } else if (state.is(Blocks.REPEATER)) {
