--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -537,6 +_,7 @@
         }
     }
     // Paper end - optimise entity tracker
+    public boolean pushedByFluid = true; // Sakura - entity pushed by fluid api
 
     // Purpur start - Add canSaveToDisk to Entity
     public boolean canSaveToDisk() {
@@ -4366,7 +_,7 @@
     }
 
     public boolean isPushedByFluid() {
-        return true;
+        return this.pushedByFluid; // Sakura - entity pushed by fluid api
     }
 
     public static double getViewScale() {
@@ -5432,7 +_,7 @@
         void accept(Entity entity, double x, double y, double z);
     }
 
-    record Movement(Vec3 from, Vec3 to, Optional<Vec3> axisDependentOriginalMovement) {
+    public record Movement(Vec3 from, Vec3 to, Optional<Vec3> axisDependentOriginalMovement) {
         public Movement(Vec3 from, Vec3 to, Vec3 axisDependentOriginalMovement) {
             this(from, to, Optional.of(axisDependentOriginalMovement));
         }
