From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Wed, 16 Aug 2023 22:34:49 +0100
Subject: [PATCH] Store Entity Data/State


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 2c5128ceed77e8d49855cc005597d3b9c804fb72..1594b96667c3a376b4071dad7bb792e911ede765 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -554,6 +554,21 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return flags;
     }
     // Sakura end - load chunks on movement
+    // Sakura start - store entity data/state
+    private me.samsuik.sakura.entity.EntityState entityState = null;
+
+    public final void storeEntityState() {
+        this.entityState = me.samsuik.sakura.entity.EntityState.of(this);
+    }
+
+    public final me.samsuik.sakura.entity.EntityState entityState() {
+        return this.entityState;
+    }
+
+    public final boolean compareState(final Entity to) {
+        return to.entityState() != null && to.entityState().comparePositionAndMotion(this);
+    }
+    // Sakura end - store entity data/state
 
     // Purpur start - Add canSaveToDisk to Entity
     public boolean canSaveToDisk() {
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 31decd76f2444e72e3d54a745b3a443829f0f4c2..583d23d99c1f5b355b2ee14896129094bec1524a 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1495,6 +1495,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
     public <T extends Entity> void guardEntityTick(Consumer<T> action, T entity) {
         try {
+            entity.storeEntityState(); // Sakura - store entity data/state
             action.accept(entity);
         } catch (Throwable var6) {
             // Paper start - Prevent block entity and entity crashes
