From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 25 Nov 2023 21:14:45 +0000
Subject: [PATCH] Allow explosions to destroy lava


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 06993303103dcd7241cbd4ecc9d24b981b2d2118..6df8775b020d329b6b4512f9476c3c3dbd08b4c5 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -404,6 +404,11 @@ public class ServerExplosion implements Explosion {
                 return Optional.of(ZERO_RESISTANCE);
             }
             // Sakura end - destroy water logged blocks
+            // Sakura start - allow explosions to destroy lava
+            if (blockState.is(net.minecraft.world.level.block.Blocks.LAVA) && this.level.sakuraConfig().cannons.explosion.explodeLava) {
+                return Optional.of(ZERO_RESISTANCE);
+            }
+            // Sakura end - allow explosions to destroy lava
         }
 
         return this.damageCalculator.getBlockExplosionResistance(this, this.level, pos, blockState, fluidState);
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 8db95b74f88f8096de93115ae8d3fb2e6184ad3b..67ee0608e6ba08b7ff9931aec8b835870b813ce0 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -201,7 +201,7 @@ public abstract class BlockBehaviour implements FeatureElement {
                 state.getDrops(builder).forEach(stack -> dropConsumer.accept(stack, pos));
             }
 
-            level.setBlock(pos, Blocks.AIR.defaultBlockState(), 3);
+            level.setBlock(pos, Blocks.AIR.defaultBlockState(), level.sakuraConfig().cannons.explosion.explodeLava && state.is(Blocks.LAVA) ? 2 : 3); // Sakura - allow explosions to destroy lava; don't cause block updates when blowing up lava
             block.wasExploded(level, pos, explosion);
         }
     }
