From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 2 Dec 2023 15:14:15 +0000
Subject: [PATCH] Legacy lava block formation


diff --git a/net/minecraft/world/level/block/LiquidBlock.java b/net/minecraft/world/level/block/LiquidBlock.java
index 5d44115aa9f23f3e8ece6c48ed888e23d23a8b77..2cd4e876de244fb2b7c7f1efe7fca8c2d3fc5c79 100644
--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -202,7 +202,14 @@ public class LiquidBlock extends Block implements BucketPickup {
                     // Sakura start - configure server mechanics
                     final FluidState fluidState = state.getFluidState();
                     final Block block = fluidState.isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
-                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    // Sakura start - legacy lava block formation
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget;
+                    if (level.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                        mechanicsTarget = me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.legacy();
+                    } else {
+                        mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    }
+                    // Sakura end - legacy lava block formation
                     if (block == Blocks.COBBLESTONE && !me.samsuik.sakura.mechanics.LegacyBlockFormation.canLiquidFormBlock(level, pos, fluidState, mechanicsTarget)) {
                         return true;
                     }
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index a57dd74afc1b37f6c50a719c70c9e554ff01c2de..09a5912a50b42144ccf6b16af8e2e1a1059eabad 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -185,9 +185,16 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(FluidState fluidState, BlockGetter blockReader, BlockPos pos, Fluid fluid, Direction direction) {
-        // Sakura start - configure server mechanics
-        if (blockReader instanceof Level level && level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
-            return false;
+        // Sakura start - configure server mechanics & legacy lava block formation
+        if (blockReader instanceof Level level) {
+            if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+                return false;
+            }
+
+            if (level.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                return false;
+            }
+            // Sakura end - legacy lava block formation
         }
         // Sakura end - configure server mechanics
         return fluidState.getHeight(blockReader, pos) >= 0.44444445F && fluid.is(FluidTags.WATER);
diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index 5140b563b98004eada72dcc9213f4ddb367d16bc..55af2c77b20f1d41814fda1d18e09391b12b05dd 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -131,10 +131,18 @@ public abstract class WaterFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(FluidState fluidState, BlockGetter blockReader, BlockPos pos, Fluid fluid, Direction direction) {
-        // Sakura start - configure server mechanics
-        final boolean canReplace = direction == Direction.DOWN
-            || blockReader instanceof Level level
-            && level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13);
+        // Sakura start - configure server mechanics & legacy lava block formation
+        boolean canReplace = false;
+        if (direction == Direction.DOWN) {
+            canReplace = true;
+        } else if (blockReader instanceof Level level) {
+            if (level.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+                canReplace = true;
+            } else if (level.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                canReplace = true;
+            }
+        }
+        // Sakura end - legacy lava block formation
         // Before 1.13 lava could replace water
         return canReplace && !fluid.is(FluidTags.WATER);
         // Sakura end - configure server mechanics
