From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Sat, 2 Dec 2023 15:14:15 +0000
Subject: [PATCH] Legacy lava block formation


diff --git a/net/minecraft/world/level/block/LiquidBlock.java b/net/minecraft/world/level/block/LiquidBlock.java
index c22c4cacadfe3c3a1291ba4a83f966e7b5e9d9af..155d9ad1e5fecb46033e0ad5e3ed2c49825e3cd8 100644
--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -204,7 +204,14 @@ public class LiquidBlock extends Block implements BucketPickup {
                     // Sakura start - configure server mechanics
                     final FluidState fluidState = state.getFluidState();
                     final Block block = fluidState.isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
-                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    // Sakura start - legacy lava block formation
+                    final me.samsuik.sakura.mechanics.MinecraftMechanicsTarget mechanicsTarget;
+                    if (level.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                        mechanicsTarget = me.samsuik.sakura.mechanics.MinecraftMechanicsTarget.legacy();
+                    } else {
+                        mechanicsTarget = level.localConfig().at(pos).mechanicsTarget;
+                    }
+                    // Sakura end - legacy lava block formation
                     if (block == Blocks.COBBLESTONE && !me.samsuik.sakura.mechanics.LegacyBlockFormation.canLiquidFormBlock(level, pos, fluidState, mechanicsTarget)) {
                         return true;
                     }
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index be63a51052a8331c6dbf6ccd4028a7873d8a2e76..6c1cdd32a99856610e85dd66d5b647667a15312c 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -185,9 +185,16 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(FluidState fluidState, BlockGetter level, BlockPos pos, Fluid fluid, Direction direction) {
-        // Sakura start - configure server mechanics
-        if (level instanceof ServerLevel serverLevel && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
-            return false;
+        // Sakura start - configure server mechanics & legacy lava block formation
+        if (level instanceof ServerLevel serverLevel) {
+            if (serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+                return false;
+            }
+
+            if (serverLevel.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                return false;
+            }
+            // Sakura end - legacy lava block formation
         }
         // Sakura end - configure server mechanics
         return fluidState.getHeight(level, pos) >= 0.44444445F && fluid.is(FluidTags.WATER);
diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index 3dc4d3606b8f1fece6305269818a0931729a02c4..69f02ce49a3a2cd3c9115246fd0eff7b8daafadb 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -130,10 +130,18 @@ public abstract class WaterFluid extends FlowingFluid {
 
     @Override
     public boolean canBeReplacedWith(FluidState fluidState, BlockGetter level, BlockPos pos, Fluid fluid, Direction direction) {
-        // Sakura start - configure server mechanics
-        final boolean canReplace = direction == Direction.DOWN
-            || level instanceof ServerLevel serverLevel
-            && serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13);
+        // Sakura start - configure server mechanics & legacy lava block formation
+        boolean canReplace = false;
+        if (direction == Direction.DOWN) {
+            canReplace = true;
+        } else if (level instanceof ServerLevel serverLevel) {
+            if (serverLevel.localConfig().at(pos).mechanicsTarget.before(me.samsuik.sakura.mechanics.MechanicVersion.v1_13)) {
+                canReplace = true;
+            } else if (serverLevel.sakuraConfig().environment.blockGeneration.legacyBlockFormation) {
+                canReplace = true;
+            }
+        }
+        // Sakura end - legacy lava block formation
         // Before 1.13 lava could replace water
         return canReplace && !fluid.is(FluidTags.WATER);
         // Sakura end - configure server mechanics
